### ISSUE TITLE
generate_series(_,_) generates error because of deficient static type checking (so generate_series() should probably be more like range())

### ISSUE THREAD

**Issue by pkoppstein** [duckdb/duckdb#14012]
### What happens?

The program below is a working prime number tester, and works even for primep(1):
```
D select primep(1);
┌───────────┐
│ primep(1) │
│  boolean  │
├───────────┤
│ false     │
└───────────┘
```
However, if one replaces the line containing the call to unnest()
by the commented-out line immediately above it, primep(1) no longer works.
It produces the message:
```
Binder Error: start is bigger than end, but increment is positive: cannot generate infinite series
```
Evidently, the guard `when nnumber > 2' is insufficient.

If it's too much to hope that the static type checking can be improved to 
prevent the problem, then I believe a perfectly fine solution would be to
make generate_series() more like range() (that is, not produce an error when
an error is not actually required).

In this connection, it's worth noting that:
```
from unnest(generate_series(2,1)) as t(n); 
```
succeeds:
```
┌────────┐
│   n    │
│ int64  │
├────────┤
│ 0 rows │
└────────┘
```
However an attempt to use this feature in the program below fails for reasons which are probably
also worth looking into.  

### To Reproduce

```

create or replace function primep(nnumber) as (
  select
    case
    when nnumber < 2 then false
    when nnumber = 2 then true
    when nnumber > 2 then
      case when exists
         ( select * from 
           ( select (nnumber % anumber) as modNumber
             -- from generate_series(2, nnumber) _(anumber)  -- This ought to work because of the 'number > 2' guard
             from (select unnest(range(2, nnumber)) as anumber)
           ) 
           where modNumber = 0 
         ) 
       then false
       else true
       end
    end
);          
select primep(1);
```

### OS:

MacOS

### DuckDB Version:

1.0, 1.1

### DuckDB Client:

CLI

### Full Name:

Peter Koppstein

### Affiliation:

Princeton University

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a source build

### Did you include all relevant data sets for reproducing the issue?

Not applicable - the reproduction does not require a data set

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by soerenwolfers** [comment]
This has nothing to do with `generate_series` vs `range` (observe that the error persists if you replace `generate_series` by `range`, and conversely the error disappears even if you replace `from generate_series` by `from (select unnest(generate_series(...)))`)

It just boils down to the fact that the "table function" (not sure of the proper terminology)

```sql
from generate_series(0, -1)
```
```
Binder Error: start is bigger than end, but increment is positive: cannot generate infinite series
```

is different from the "array function" (not sure of the proper terminology)

```sql
select generate_series(0, -1)
```
```
┌────────────────────────┐
│ generate_series(0, -1) │
│        int64[]         │
├────────────────────────┤
│ []                     │
└────────────────────────┘
```

I'd tend to agree with you that asking for the two functions to use the same semantics regarding the start and endpoints is more of a bug report than a feature request.

**Comment by pkoppstein** [comment]
@soerenwolfers - Thanks for pointing out the error in my attempted diagnosis. It would be good to know if there was a reason for the discrepancy between the behavior of generate_series() and range().  I hope you agree, though, that the current behavior is incorrect - either the outermost CASE statement should be sufficient to prevent the error; or the attempted "compile-time" check should be dropped altogether.

**Comment by soerenwolfers** [comment]
> It would be good to know if there was a reason for the discrepancy between the behavior of generate_series() and range()

again, there is no difference between `generate_series` and `range` wrt to the behavior at issue here. The only difference between the two, which your OP actually gets wrong (although it doesn't matter for looking for divisors) is that range excludes the right end point. 

> or the attempted "compile-time" check should be dropped altogether

I agree that the "table function" `generate_series` shouldn't be invoked unless necessary. This should be fixable indeed, since It already isn't invoked when the endpoints are "dynamic", e.g. 
```sql
SELECT 
CASE WHEN x > 0 THEN (FROM range(0, x)) END 
FROM (SELECT -1 AS x)
```
works but 
```sql
SELECT 
CASE WHEN x > 0 THEN (FROM range(0, -1)) END 
FROM (SELECT -1 AS x)
```
throws.

**Comment by soerenwolfers** [comment]
> It would be good to know if there was a reason for the discrepancy between the behavior of generate_series() and range()

again, there is no difference between `generate_series` and `range` wrt to the behavior at issue here. The only difference between the two, which your OP actually gets wrong (although it doesn't matter for looking for divisors) is that range excludes the right end point. 

> or the attempted "compile-time" check should be dropped altogether

I agree that the "table function" `generate_series` in your original example shouldn't be invoked. This is really not a matter of compile checks, it's just a missed optimization opportunity. For example, 
```sql
SELECT CASE WHEN 0 > 1 THEN (FROM range(0, 100000000000) select sum(range)) END 
```
doesn't finish in human time, but if it did it would just return `NULL`.

Whether this missed optimization opportunity is actually a bug (because it could allow for side-effects where none are supposed to occur, such as for example the side effect of an error being thrown on invalid inputs) is something that someone more familiar with duckdb/SQL will have to answer. 


---


In summary, I believe there are two separate issues here:

(1) The `range` & `generate_series` functions both have "array" and "table" versions, that have different semantics wrt to constraints on the endpoints. This is not a bug, because they *are* different things (for comparison, `array_value` is a regular "array returning function" but not a "table function" that can be used in a `from` clause). Nonetheless, I would guess it's not desired behavior and should be fixed.

(2) duckdb executes subqueries in arms of CASE clauses even when those arms don't end up being visited. This is definitely a missed optimization opportunity (see example above that takes forever despite just returning NULL), but again I'm not 100% that it counts as a bug.

**Comment by pkoppstein** [comment]
In case it's not clear, the main reason for wanting to be able to write `from generate_series()` here (as opposed to `select unnest()`) is that the former is many times faster.

To give an idea of how much fast, consider the test case:
```
select primep( 999966000289 );  -- 999983 * 999983
```

Using the lines:
```
generate_series(2, 1 + sqrt(nnumber)::BIGINT) t(anumber)
```
or
```
(select unnest(range(2, 1 + sqrt(nnumber)::BIGINT)) as anumber)
```
I measured the respective u+s times (including invocation, for which u+s = 0.026) as: 
```
series:  0.063
unnest:  0.377
```

**Comment by soerenwolfers** [comment]
still an issue

**Comment by soerenwolfers** [comment]
@szarnyasg Sorry for accidentally keeping this alive; my last comment is wrong, this issue has been fixed so it can be closed:

```sql
from generate_series(0, -1)
```
now returns no rows instead of throwing, which matches the behavior of `select generate_series(0, -1)`. 

(I got confused because the snippet in the OP that wasn't working isn't producing the semantically correct results but that's just because of an independent off by one bug)
